### DISCUSSION TITLE
Table Sparklines and Bars

### DISCUSSION THREAD

**Discussion by mcrascal** [evidence-dev/evidence#2370] [General]
Interface for adding sparklines and bars to data tables. 

See discussion and inspiration here #498

**Comment by hughess** [comment]
Here are some examples of these:

### Bars
```html
 <Column id=value_usd contentType=bar/>
```

![table-barchart](https://github.com/evidence-dev/evidence/assets/12602440/f435de7d-68e9-4b5a-bc28-a9dfeb3787fa)

### Sparklines / sparkbars
```html
 <Column id=value_usd contentType=sparkbar/>
```

![table-sparkline](https://github.com/evidence-dev/evidence/assets/12602440/47e00bd3-0899-4338-b63c-eedb81ffe478)


And one other idea:

### Dot
```html
 <Column id=value_usd contentType=dotline/>
```

<img width="580" alt="dots" src="https://github.com/evidence-dev/evidence/assets/12602440/efaa6132-94a7-4749-9031-6f54f7741992">

**Comment by archiewood** [comment]
Beautiful. Are these Echarts?

**Comment by hughess** [comment]
Nope! The bars and dots are custom svg, and the sparkbar one is svelte-tiny-charts (we'd probably want to switch to echarts or another library for that one to avoid the dependency issue from that library). 

Here's the code for the bar one (but with different colours I was trying out):

```svelte
<svg width="100%" height="{column.height ?? '1.5em'}" style="padding-top: 5px;">
    <rect width="{row[column.id] / (column.barMax ?? Math.max(...data.map(d => d[column.id]))) * 100}%" height="100%" fill=navy opacity=1></rect>
    <text x="5%" y="60%" fill=white>{formatValue(row[column.id], safeExtractColumn(column).format)}</text>                    
</svg>
```

And for the dot line:

```svelte
<svg width=60% height=3em>
     <line x1="0" y1="50%" x2="100%" y2="50%" stroke="lightgrey" stroke-dasharray=2 />
     <circle cx="{Math.min(row[column.id] / Math.max(...data.map(d => d[column.id])) * 100, 95)}%" cy="50%" r="4" fill=navy></circle>
</svg>
```

**Comment by mcrascal** [comment]
Nice -- the dots are my favourite. How do you indicate the x axis for the sparkbar?

**Comment by hughess** [comment]
The way I did it above was accept another dataset into the `Column` component:

```html
<DataTable data={state_data}>
    <Column id=state/>
    <Column id=current_value/>
    <Column id=current_value title="Trend" align=center contentType=sparkline sparklineData={sparkline_data}/>
</DataTable>
```

I don't fully remember, but I think I also think I pulled the sparkline logic from BigValue into a Sparkline component and set it up to run on any dataset with a date column.

Then I added this loop into the DataTable code to check whether each row in the `sparkLineData` dataset matched the current row in the `DataTable`. For example, if you included a `state` column and a `year` column in your `sparklineData`, this would only show a sparkline on rows in the data table where both of those columns are matched:

```svelte
 <Sparkline data={column.sparklineData.filter(d => {
      let status = true;
      if(column.otherCols){
          for(let i=0; i<column.otherCols.length; i++){
                 if(d[column.otherCols[i]] === row[column.otherCols[i]]){
                      status = status * true
                  } else {
                       status = false
                  }
           }
     }
    return status;
})}/>
```

**Comment by hughess** [comment]
https://github.com/evidence-dev/evidence/pull/2608

**Comment by hughess** [comment]
Closed by #2608
